### DISCUSSION TITLE
why does the L0Cache controller also need to be connected to the router in ruby cache?

### DISCUSSION THREAD

**Discussion by qqli579** [gem5/gem5#567] [Q&A] [answered]
I am now using the file /configs/ruby/MESI_Three_Level.py to build a Mesh_XY 4x4 mesh network. I learned that L0Cache and L1Cache in this file are private to the CPU, and L2Cache is shared. But why do we need to connect the L0Cache controller to the router in the create_topology function? I think we only need the L1Cache and L2Cache controllers to connect to the router.
If the controller of L0Cache is connected to the router, does L0Cache become shared?

**Comment by powerjg** [comment, marked as answer]
There must be a way to deliver messages from one controller to another (the L0 to the L1 cache controllers in this case). The only way messages can move between message buffers is via the network. To move messages via the network you need a router. This is simply how Ruby is designed.

That said, there is a special interconnect object called a "wire buffer" that *does not* require a router. This may be able to be used between the L0 and the L1, though I've never seen it used there.

**Comment by lsj134679** [comment]
I would like to consult you on a question. Currently, I want to implement an RISC-V architecture with a cache coherence protocol of MESI_Two_Level, and a network topology of mesh architecture, in fs mode. Can the gem5 simulator achieve this?
